ui-manchette: fix waypoint menu positioning

SharglutDev opened the pull request:
- Move the menu next to its related waypoint so it can always be positioned relatively to it
- To keep osrd-ui as neutral as possible, the menu has no styles, they need to be passed by props

fix #667 

After discussion with @thibautsailly, it has been decided to totally block the scroll in the manchette (and only in the manchette) as long as the menu is displayed. If we still try to scroll, the whole page will scroll.

SharglutDev committed:
ui-manchette: fix waypoint menu positioning

- Move the menu next to its related waypoint so it can always be positioned relatively to it
- To keep osrd-ui as neutral as possible, the menu has no styles, they need to be passed by props
- Add a custom hook using the intersection observer api. This will allow the menu to hide when its waypoint is hidden on scroll
- This hook will remain in osrd-ui to prevent a user to have the implement this necessary logic when adding a waypoint menu to the manchette

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on ui-manchette/src/components/OperationalPoint.tsx:
```diff
@@ -27,7 +32,7 @@ const OperationalPoint = ({
       ref={opRef}
       onClick={() => {
-        if (onClick) onClick(id, opRef.current);
+        if (onClick) onClick(`${id}-${position}`); // to handle waypoints with same id
```
Hm, passing in here a custom string which isn't an ID seems pretty surprising… Especially since the argument is named `opId` in the type definition.

But why do we need to pass any argument here? The callers already provide a callback per-OP, no?

SharglutDev replied:
Hmm but if we don't pass any arguments, how can we have the information of which waypoint has been clicked ?

emersion replied:
Ah, apologies, I mixed things up with the menu item callback!

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -51,7 +51,19 @@ export type OperationalPointType = ArrayElement<PathProperties['operational_poin
+export type WaypointMenuData = {
+  activeOperationalPointId?: string;
+  waypointMenuItems?: WaypointMenuItem[];
+  waypointMenuClassName?: string;
```
Do we really need to allow callers to customize the class name? Wouldn't the menu already have the correct size, since it's now styled in osrd-ui? If callers really want to customize something, can't they use a regular CSS selector?

SharglutDev replied:
My understanding was that we want to keep the manchette/spacetime as neutral as possible. That's why I went on this path. If the team using this package wants to use the waypoint menu it can use its custom style for it to match the rest of their app design.
The `.menu` class has been added here only for the related story.
Furthermore, if the team using the menu doesn't want to keep any of the current style, they will need to invalidate all of them even if it's a css property that they may not have used. It seems pretty painful to me.

Should we have another discussion for that ?

emersion replied:
Right, I'm not really getting why the manchette's styling would be in osrd-ui but the manchette menu's styling could not be. I must admit, I don't see why someone would want to change the menu's appearance, because it would then become inconsistent with the manchette.

emersion replied:
After discussion, it seems like the consensus is to pass the whole menu component from OSRD.

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -51,7 +51,19 @@ export type OperationalPointType = ArrayElement<PathProperties['operational_poin
+};
+
+export type WaypointMenuData = {
+  activeOperationalPointId?: string;
```
Same comment as https://github.com/OpenRailAssociation/osrd-ui/pull/668/files#r1816147595

Could we maybe use an index instead of a custom string to refer to a particular OP in the list?

SharglutDev replied:
Was afraid that when a waypoint is hidden when zooming out or when we hide one when clicking on it, it will break something but you're right it should be ok.

emersion replied:
Ah right, that's a good thing to keep in mind indeed! Maybe at some point we should introduce a unique waypoint ID (which would be different if the same OP appears twice), similar to OSRD's `PathItemLocation.id`?

But TBH, why does osrd-ui even need to access the OP ID? Couldn't we always pass a guaranteed unique waypoint ID in there? I suppose all of these questions are related to https://github.com/OpenRailAssociation/osrd-ui/issues/645.

emersion commented on ui-manchette/src/styles/manchette.css:
```diff
@@ -1,5 +1,44 @@
+        border-bottom: 0.0625rem solid #d3d1cf;
+      }
+
+      /* .icon { */
+      span:first-child {
```
Should we use the class name here instead? Any reason why `:first-child` is used?

SharglutDev replied:
Because I went to the full customize option for the waypoint menu (see other comment). This would mean the user needs to pass more props for each `className` of the menu, it seemed pretty complicated to me.

emersion replied:
Hm, that sounds like a somewhat awkward middle ground to me: on one hand the structure of the menu is entirely managed by osrd-ui, on the other hand all of the styling is performed outside.

What's the use-case for keeping the styling outside? The menu appearance needs to be consistent with the manchette's styling, so I'm not sure I understand the motivation here?

If we really want to keep the styling out of osrd-ui, why do we need to have the menu and menu item components inside osrd-ui? We could pass the menu component from the outside as a prop.

SharglutDev replied:
I honestly don't have any passion to keep this at it is now. I would even agree that it would be nice for a team to have a fully functional waypoint menu if they wanted it. I just went that way because for my understanding it had been decided that the spacetime/manchette had to remain as neutral as possible.

But I agree with you that right now we are in a middle ground :)

emersion commented on ui-manchette/src/components/WaypointMenu.tsx:
```diff
@@ -0,0 +1,33 @@
+  <div className={className}>
+    {items.map(({ title, icon, onClick }) => (
+      <button
+        key={title}
```
Human-readable strings aren't supposed to be used as keys in general, because they're not guaranteed to be unique.

SharglutDev replied:
Good point. Used `index` as the menu items won't change.

emersion commented on ui-manchette/src/components/WaypointMenu.tsx:
```diff
@@ -0,0 +1,33 @@
+      <button
+        key={index}
+        type="button"
+        onClick={(e) => {
+          onClick(e);
+        }}
```
Nit: can be simplified to `onClick={onClick}`

SharglutDev replied:
Oh I did it again 🫣

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -51,7 +51,12 @@ export type OperationalPointType = ArrayElement<PathProperties['operational_poin
+
+export type WaypointMenuData = {
+  menu: React.ReactNode;
+  activeWaypointIndex?: number;
```
Once #651 is merged, we'll be able to switch back to a waypoint ID here if we want to (and in the `onClick` callback)

SharglutDev replied:
Ok so do I leave it like that and it will be dealt with in #651 or do we want to wait for it to be done ?

emersion replied:
I suppose it depends in which order the PRs get merged…

emersion commented on ui-manchette/src/styles/manchette.css:
```diff
@@ -1,5 +1,6 @@
 .manchette-container {
   .manchette-actions {
+    z-index: 11;
```
Not 100% sure we want to place the menu behind the toolbar… Wouldn't this always hide the menu for the very last waypoint?

![out](https://github.com/user-attachments/assets/0dd2e96d-d593-4f6d-95a3-1d708cd8e2d5)

SharglutDev replied:
It was a request from @Tguisnet but you're right we will have an issue for the last waypoints !

emersion reviewed (approved):
LGTM apart from this last question!

emersion commented on ui-manchette-with-spacetimechart/src/stories/useElementInView.ts:
```diff
@@ -0,0 +1,48 @@
+      }
+    );
+
+    observer.observe(currentRef.parentElement!);
```
In the `Waypoint` component the ref is moved to a child, and here we grab the parent: could we just leave the ref on the parent in `Waypoint` instead?

But IIRC watching the parent causes issues in proportional mode when scrolling up?

SharglutDev replied:
Didn't we check that if the ref was on the parent, the menu was not displayed as expected because the parent height is large when zoomed in ?

emersion replied:
Hm, not sure I understand…

I think we should keep passing the same element to `observer.observe()`, but I don't think we should be passing `parentElement`. Instead, we should make it the caller's responsibility to pass a ref targeting the parent element. Does that make sense?

emersion commented on ui-manchette-with-spacetimechart/src/stories/useElementInView.ts:
```diff
@@ -0,0 +1,48 @@
+  containerRef?: RefObject<HTMLElement>,
+  options: IntersectionObserverInit = {
+    threshold: 1,
+    rootMargin: `0px 0px ${MANCHETTE_ACTIONS_HEIGHT} 0px`,
```
This hooks sounds like it's generic, but then we're hardcoding details specific to a particular use-cases in the default props. Should we move these defaults to the caller site instead?

SharglutDev replied:
Totally agreed, done.

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -10,7 +10,13 @@ export type Waypoint = {
+export type WaypointMenuData = {
+  menu: React.ReactNode;
+  activeWaypointId?: string;
+  isWaypointInView?: boolean;
```
Instead of adding a new `isWaypointInView` field, can't the caller pass a null `menu` field instead? (I believe `React.ReactNode` allows for `null`.)

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -10,7 +10,13 @@ export type Waypoint = {
   styles?: CSSProperties;
   display?: boolean;
-  onClick?: (opId: string, opRef: HTMLDivElement | null) => void;
+  onClick?: (waypointId: string, waypointRef: React.RefObject<HTMLDivElement>) => void;
```
Note to self: I'm a bit sad to have the caller take care of figuring out whether the waypoint is visible or not, it would be nicer API. Plus the underlying `HTMLDivElement` passed here to the callback can become stale while the menu is displayed (React can choose to re-render it in another DOM element any time).

 But IIRC this causes other Issues™.

SharglutDev replied:
As discussed, gonna keep the menu visibility in osrd-ui and the caller will only be responsible to pass the container ref to the manchette if he want's to have the hiding behavior.

emersion commented on ui-manchette/src/components/Waypoint.tsx:
```diff
@@ -9,13 +9,18 @@ import { positionMmToKm } from '../utils';
   waypoint: InteractiveWaypoint;
   isActive: boolean;
+  waypointMenu?: React.ReactNode;
+  waypointsContainerRef: React.RefObject<HTMLDivElement>;
```
Sounds like this prop is unused?

SharglutDev replied:
Right, done

SharglutDev committed:
ui-manchette: update some package exports

- Add some type export needed when implementing the app
- Remove some exports not needed outside of ui-manchette

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on storybook/.storybook/main.ts:
```diff
@@ -24,13 +24,13 @@ const config: StorybookConfig = {
-      },
-    });
-  },
+  // async viteFinal(config) {
+  //   return mergeConfig(config, {
+  //     resolve: {
+  //       preserveSymlinks: true,
+  //     },
+  //   });
+  // },
```
Note: this should be left out of this PR, since these changes are also here: https://github.com/OpenRailAssociation/osrd-ui/pull/721

SharglutDev replied:
Oups was to hurry to eat the couscous :x

emersion commented on ui-manchette-with-spacetimechart/src/stories/base.stories.tsx:
```diff
@@ -1,7 +1,6 @@
 
-import { Manchette } from '@osrd-project/ui-manchette';
+import Manchette from '@osrd-project/ui-manchette';
 import type { ProjectPathTrainResult, Waypoint } from '@osrd-project/ui-manchette/dist/types';
```
Now that these are exported from `@osrd-project/ui-manchette` (without `dist/`), maybe we could use that instead?

SharglutDev replied:
Good point, done

emersion reviewed (commented):
Looks good apart from these remaining comments :)

SharglutDev committed:
ui-manchette: fix manchette toolbar not fixed bottom

The PR https://github.com/OpenRailAssociation/osrd-ui/pull/728 changed the height of the manchette from a min height to a fixed one.
It introduced a bug where, when having a long list of waypoint, when scrolling bottom, the manchette toolbar wasn't stick to the bottom anymore.
Fix this issue by giving back a min height to the manchette while still keeping its dynamic value introduced in PR #728.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

anisometropie commented on ui-manchette/src/hooks/useElementInView.ts:
```diff
@@ -0,0 +1,50 @@
+    const root = containerRef?.current || null;
+
+    if (!currentRef) {
+      // If the user implementing ui-manchette doesn't provide a ref to handle the menu visibility,
```
Why would one use this hook and not provide a ref ?

I guess it would also be the case if the component is not yet loaded, why set isInView as true in that case?

SharglutDev replied:
Like we discussed, it was a leftover from the old way I used this hook. I updated the comment to match the new logic.

SharglutDev replied:
Because of design changes, the hook has been removed.

anisometropie commented on ui-manchette/src/hooks/useElementInView.ts:
```diff
@@ -0,0 +1,50 @@
+import { useState, useEffect, type RefObject } from 'react';
```
This seems we could reuse this hook for general purpose, why restrain it to ui-manchette only ?

anisometropie replied:
like move it to ui-core?

SharglutDev replied:
As discussed, I think we can leave it here for now and if it's needed in another subrepo, we can move it at this moment.

SharglutDev replied:
See previous comment.

anisometropie commented on ui-manchette/src/hooks/useElementInView.ts:
```diff
@@ -0,0 +1,50 @@
+    const root = containerRef?.current || null;
+
+    if (!currentRef) {
+      // If the user implementing ui-manchette doesn't provide a ref to handle the menu visibility,
```
if it’s a generic component it shouldn’t mention manchette specific behavior

SharglutDev replied:
Same as previous comment. It's only related to the manchette for now

SharglutDev replied:
See previous comment.

anisometropie reviewed (changes requested):
the behavior is not consistent.
open the menu, then scroll down, and click and the menu shows on top of the waypoint the next time
https://github.com/user-attachments/assets/d5762a6b-62e8-4454-8fe9-455157bdc641

SharglutDev commented:
After discussion with @thibautsailly, it has been decided to totally block the scroll in the manchette (and only in the manchette) as long as the menu is displayed. If we still try to scroll, the whole page will scroll. (reported in the PR and issue description)

emersion commented on ui-manchette/src/components/Waypoint.tsx:
```diff
@@ -6,42 +6,57 @@ import { type InteractiveWaypoint } from '../types';
+        {waypointMenu && isActive && waypointRef.current && (
+          <div
+            className="menu-wrapper"
+            style={{ top: waypointRef.current.getBoundingClientRect().bottom - MENU_OFFSET }}
```
This performs some layout-related computations in the rendering function. This is unreliable, since at rendering time the DOM hasn't been updated yet.

emersion commented on ui-manchette/src/styles/waypoint.css:
```diff
@@ -81,12 +80,22 @@
+
+.last-separator {
+  width: 26px;
+  height: 2rem;
```
What is the purpose of this?

SharglutDev replied:
Was needed when removing the position relative for the waypoint. It is not needed anymore with the new logic and has been reverted.

emersion commented on ui-manchette/src/components/Waypoint.tsx:
```diff
@@ -6,42 +6,57 @@ import { type InteractiveWaypoint } from '../types';
+        {waypointMenu && isActive && waypointRef.current && (
+          <div
+            className="menu-wrapper"
+            style={{ top: waypointRef.current.getBoundingClientRect().bottom - MENU_OFFSET }}
```
This sets a position relative to the viewport, but this breaks as soon as a parent is set as `position: relative`. Since the manchette is a re-usable component, it shouldn't assume anything about the positioning mode of its parent.

I believe this won't work in OSRD since `.chart-container` has `position: relative`.

SharglutDev replied:
You're totally right. Logic has been changed and the menu is now positioned relatively to the manchette/spacetime wrapper.

emersion commented:
This approach still sounds pretty fragile to me.

- If we really want to block interactions while the menu is opened, I'd suggest [using a portal](https://react.dev/reference/react-dom/createPortal#rendering-to-a-different-part-of-the-dom) to insert a fullscreen transparent overlay preventing all interactions except for the menu, and positioning the menu inside that overlay.
- If we want to allow interactions while the menu is opened, I'd suggest we insert the menu outside of the manchette via a portal, then use `requestAnimationFrame` to update the position of the menu dynamically if the waypoint is visible.

emersion commented on ui-manchette/src/components/Manchette.tsx:
```diff
@@ -26,40 +29,82 @@ const Manchette = ({
+    if (!manchetteWrapperPosition || !waypointPosition) {
+      setMenuPosition(undefined);
+    } else {
+      setMenuPosition(window.scrollY * 2 + manchetteWrapperPosition + waypointPosition);
```
Hm, not sure how this works, and seems like this doesn't work on my machine… Why multiply the window scroll offset by 2?

The menu has `position: absolute`, and it's inside a wrapper which has `position: relative`. This means we need to compute the distance between these two elements. `waypointPosition - manchetteWrapperPosition` worked well for me.

SharglutDev replied:
Oh right, I was so focused on trying to handle the scrollY that I didn't think to do it that way.

emersion commented on ui-manchette/src/components/Manchette.tsx:
```diff
@@ -26,40 +29,82 @@ const Manchette = ({
+  const [menuPosition, setMenuPosition] = useState<number>();
+
+  // Allow to track the menu position after we might have scrolled in the page
+  useEffect(() => {
```
We measure DOM elements from inside this function, so we should use `useLayoutEffect()` instead, so that our callback isn't called before the DOM is updated.

SharglutDev replied:
Still got to keep this one in mind, thank you

emersion commented on ui-manchette/src/components/Manchette.tsx:
```diff
@@ -26,40 +29,82 @@ const Manchette = ({
+
+  return (
+    <div className="manchette-container">
+      {waypointMenuData?.menu && menuPosition && (
```
Nit: would be safer to check for `menuPosition !== undefined`, because 0 is a valid position.

SharglutDev replied:
Good point, done

emersion commented on ui-manchette/src/types.ts:
```diff
@@ -10,7 +10,14 @@ export type Waypoint = {
+export type WaypointMenuData = {
+  menu: React.ReactNode;
+  activeWaypointId?: string;
+  activeWaypointRef?: React.RefObject<HTMLDivElement>;
```
It's a bit cumbersome for the library user to grab the active waypoint ref from the `onClick` callback, then pass it back as a prop to the library. It would be nicer if this ref was entirely managed by the library itself.

To do so, we can make it so `Manchette` creates the ref, and passes it down to the active `Waypoint`. Basically,  we can:

- In `Waypoint`, add a new `waypointRef?: React.Ref<HTMLDivElement>` property
- In `WaypointList`, add a new `activeWaypointRef?: React.Ref<HTMLDivElement>` property and pass `waypointRef={activeWaypointId === waypoint.id ? activeWaypointRef : null}`
- In `Manchette`, create the `activeWaypointRef` ref and pass it to `WaypointList`

And then we can drop `activeWaypointRef` from the story/callback/etc.

SharglutDev replied:
You're right it's better ! Done

emersion commented on ui-manchette/src/components/Manchette.tsx:
```diff
@@ -1,15 +1,18 @@
-import type { InteractiveWaypoint } from '../types';
+import type { InteractiveWaypoint, WaypointMenuData } from '../types';
+
+const WAYPOINT_MENU_LEFT_OFFSET = 8;
+const WAYPOINT_MENU_TOP_OFFSET = -2;
```
Instead of having these consts, it would be nicer to move these into CSS. For instance:

```css
.menu-wrapper {
  left: 8px;
  margin-top: 30px;
}
```

SharglutDev replied:
Done

emersion reviewed (commented):
We're getting close! A few new comments.

emersion reviewed (approved):
LGTM, thanks a lot for your perseverance through this odyssey!

anisometropie reviewed (approved):
tested again, Good job !

anisometropie reviewed (commented):
just a bit unsure about disabling completely all the interactions outside of the menu.
it’s pretty natural to close the menu by clicking outside of it

SharglutDev commented:
> just a bit unsure about disabling completely all the interactions outside of the menu. it’s pretty natural to close the menu by clicking outside of it

It's just the story working like that. In osrd, you will be able to close the menu when clicking outside of it of course.
